### ISSUE TITLE
[Question] How to analyze a whole project with ikos-scan and cmake?

### ISSUE THREAD

**Issue by vrnithinkumar** [NASA-SW-VnV/ikos#151]
I was trying to analyze a whole project with ikos-scan following the [documentation](https://github.com/NASA-SW-VnV/ikos/blob/master/analyzer/README.md#analyze-a-whole-project-with-ikos-scan) and [issue 25](https://github.com/NASA-SW-VnV/ikos/issues/25). 
Unfortunately I am getting "Nothing to analyze" output from `ikos-scan make`. I was running `ikos-scan cmake [parameters] ..` and then `ikos-scan make` using generated make file.

While running `ikos-scan cmake [parameters] ..` compiler is identifying properly.
> -- Check for working CXX compiler: /ikos/ikos-install/ikos-2.1/bin/ikos-scan-c++
> -- Check for working CXX compiler: /ikos/ikos-install/ikos-2.1/bin/ikos-scan-c++ -- works

- Is this the right way to analyze a whole project with cmake?
- Is there any other documentation about ikos-scan?

**Comment by arthaud** [comment]
Hi @vrnithinkumar,

Yes, this is the right way.
No, there is currently no other documentation about ikos-scan.

Since CMake uses `ikos-scan-c++` to compile, this should work.

Are you trying to analyze a library? Please note that `ikos-scan` will only analyze binaries. To analyze a library, the easiest way is to create a small program that uses your library.

If you are trying to analyze a binary, then please try `ikos-scan -vvv make`. This should help find the problem.

By the way, if you are trying to analyze C++, see #81 and #88.

**Comment by vrnithinkumar** [comment]
Thanks @arthaud  for your reply.

I am trying to analyze the binary. The cmake builds multiple binary including binaries for the tests. 
Running with  `-vvv` option and checking a small cmake dummy project I found that `.bc` files are properly generated. But it is not linked together to a single `.bc` file for that executable. 

Looking into the logs and  `ikos/analyzer/python/ikos/scan.py `, I am suspecting [extract_bitcode(parser.output_file, bc_path) ](https://github.com/NASA-SW-VnV/ikos/blob/ea1f5905202ef6cb6644fe6585ed6a2163d4a250/analyzer/python/ikos/scan.py#L721)is not getting called for any of the binary files.

Could you please let me know how it checks for executable at link time in [scan.py:719](https://github.com/NASA-SW-VnV/ikos/blob/ea1f5905202ef6cb6644fe6585ed6a2163d4a250/analyzer/python/ikos/scan.py#L719)?
`if parser.is_link and u'executable' in filetype(parser.output_file):`

**Comment by arthaud** [comment]
First, make sure that `ikos-scan-cc` or `ikos-scan-c++` is called to link the binaries.

https://github.com/NASA-SW-VnV/ikos/blob/ea1f5905202ef6cb6644fe6585ed6a2163d4a250/analyzer/python/ikos/scan.py#L719
`parser.is_link` is True if `-c`, `-w` and `-W` are NOT present in the command line.
`filetype(parser.output_file)` is the output of the `file --brief <output_file>` command.

Could you send me the output of `ikos-scan -vvv make VERBOSE=1`?

**Comment by vrnithinkumar** [comment]
@arthaud, Once again thanks for your helpful informations.

I was able to make it work by commenting out below lines https://github.com/NASA-SW-VnV/ikos/blob/ea1f5905202ef6cb6644fe6585ed6a2163d4a250/analyzer/python/ikos/scan.py#L100-L101

From the output of the `ikos-scan -vvv`, found that the `ikos-scan-c++` is called to link the binaries.
Linking command without `-c` but it was using `-W` (`/ikos/ikos-install/ikos-2.1/bin/ikos-scan-c++   -W`). Once I commented out the lines above mentioned it was working fine.

Why is the `_set_compile_only` called in case of `-w` and `-W` present in the command line?
